File: csharp/SwiftSend/UpdateService.cs
using System.Diagnostics;
using System.Net.Http.Headers;
using System.Text.Json;

namespace SwiftSend;

internal sealed class UpdateStatus
{
    public string Current { get; init; } = "0.0.0-dev";
    public string? Latest { get; set; }
    public bool Available { get; set; }
    public string? DownloadUrl { get; set; }
    public string? AssetName { get; set; }
    public bool Checked { get; set; }
    public string? Error { get; set; }
    public bool Applying { get; set; }

    public Dictionary<string, object?> ToJson() => new()
    {
        ["current"] = Current,
        ["latest"] = Latest,
        ["available"] = Available,
        ["download_url"] = DownloadUrl,
        ["asset_name"] = AssetName,
        ["checked"] = Checked,
        ["error"] = Error,
        ["applying"] = Applying,
    };
}

internal sealed class UpdateService
{
    public const string GitHubRepo = "GuilhermeRoesler/SwiftSend";
    public static readonly string ApiLatest = $"https://api.github.com/repos/{GitHubRepo}/releases/latest";
    private const string UserAgent = "SwiftSend-Updater";
    private const int ForceKillTimeoutSec = 10;

    private readonly object _lock = new();
    private readonly UpdateStatus _status;
    private readonly string _scriptsDir;
    private readonly string _downloadDir;
    private readonly Func<string, Task<JsonDocument>> _fetchJson;
    private readonly HttpClient _http;

    public UpdateService(
        string currentVersion,
        string scriptsDir,
        Func<string, Task<JsonDocument>>? fetchJson = null,
        string? downloadDir = null)
    {
        _status = new UpdateStatus { Current = currentVersion };
        _scriptsDir = scriptsDir;
        _downloadDir = downloadDir ?? Path.Combine(Path.GetTempPath(), "SwiftSendUpdates");
        _http = new HttpClient { Timeout = TimeSpan.FromMinutes(10) };
        _http.DefaultRequestHeaders.UserAgent.ParseAdd(UserAgent);
        _http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
        _fetchJson = fetchJson ?? DefaultFetchJsonAsync;
    }

    public UpdateStatus Snapshot()
    {
        lock (_lock)
        {
            return new UpdateStatus
            {
                Current = _status.Current,
                Latest = _status.Latest,
                Available = _status.Available,
                DownloadUrl = _status.DownloadUrl,
                AssetName = _status.AssetName,
                Checked = _status.Checked,
                Error = _status.Error,
                Applying = _status.Applying,
            };
        }
    }

    public void StartBackgroundCheck()
    {
        _ = Task.Run(async () =>
        {
            try
            {
                await CheckNowAsync().ConfigureAwait(false);
            }
            catch
            {
                // ignore — status already records error
            }
        });
    }

    public async Task<UpdateStatus> CheckNowAsync()
    {
        try
        {
            using var doc = await _fetchJson(ApiLatest).ConfigureAwait(false);
            var root = doc.RootElement;
            var tag = root.TryGetProperty("tag_name", out var tagEl) ? tagEl.GetString()?.Trim() ?? "" : "";
            if (string.IsNullOrWhiteSpace(tag))
                throw new InvalidOperationException("Release sem tag_name");

            string? downloadUrl = null;
            string? assetName = null;
            if (root.TryGetProperty("assets", out var assets) && assets.ValueKind == JsonValueKind.Array)
            {
                foreach (var asset in assets.EnumerateArray())
                {
                    var name = asset.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "";
                    var url = asset.TryGetProperty("browser_download_url", out var u) ? u.GetString() ?? "" : "";
                    if (!string.IsNullOrWhiteSpace(name) && !string.IsNullOrWhiteSpace(url) && MatchesPlatform(name))
                    {
                        downloadUrl = url;
                        assetName = name;
                        break;
                    }
                }
            }

            var latestDisplay = tag.StartsWith('v') || tag.StartsWith('V') ? tag[1..] : tag;
            var newer = IsNewer(tag, _status.Current);
            var available = newer && downloadUrl is not null;

            lock (_lock)
            {
                _status.Latest = latestDisplay;
                _status.Available = available;
                _status.DownloadUrl = downloadUrl;
                _status.AssetName = assetName;
                _status.Checked = true;
                _status.Error = newer && downloadUrl is null ? "Sem artefato para esta plataforma" : null;
                return SnapshotUnlocked();
            }
        }
        catch (Exception ex)
        {
            lock (_lock)
            {
                _status.Checked = true;
                _status.Available = false;
                _status.Error = ex.Message;
                return SnapshotUnlocked();
            }
        }
    }

    public async Task<Dictionary<string, object?>> ApplyUpdateAsync()
    {
        string url;
        string assetName;
        lock (_lock)
        {
            if (_status.Applying)
                return new Dictionary<string, object?> { ["success"] = false, ["error"] = "Atualização já em andamento" };
            if (!_status.Available || string.IsNullOrWhiteSpace(_status.DownloadUrl) || string.IsNullOrWhiteSpace(_status.AssetName))
                return new Dictionary<string, object?> { ["success"] = false, ["error"] = "Nenhuma atualização disponível" };
            _status.Applying = true;
            url = _status.DownloadUrl!;
            assetName = _status.AssetName!;
        }

        try
        {
            var installer = await DownloadAsync(url, assetName).ConfigureAwait(false);
            SpawnApplyScript(installer);
            return new Dictionary<string, object?> { ["success"] = true, ["installer"] = installer };
        }
        catch (Exception ex)
        {
            lock (_lock)
            {
                _status.Applying = false;
            }
            return new Dictionary<string, object?> { ["success"] = false, ["error"] = ex.Message };
        }
    }

    public static string ReadVersion(params string[] searchRoots)
    {
        foreach (var root in searchRoots)
        {
            if (string.IsNullOrWhiteSpace(root))
                continue;
            var candidate = Path.Combine(root, "VERSION");
            if (File.Exists(candidate))
            {
                var text = File.ReadAllText(candidate).Trim();
                if (!string.IsNullOrWhiteSpace(text))
                    return text;
            }
        }

        return "0.0.0-dev";
    }

    public static (int Major, int Minor, int Patch) NormalizeVersion(string tag)
    {
        var raw = (tag ?? "").Trim();
        if (raw.StartsWith('v') || raw.StartsWith('V'))
            raw = raw[1..];
        var core = raw.Split('-', 2)[0].Split('+', 2)[0];
        var parts = new List<int>();
        foreach (var piece in core.Split('.'))
        {
            parts.Add(int.TryParse(piece, out var n) ? n : 0);
        }
        while (parts.Count < 3)
            parts.Add(0);
        return (parts[0], parts[1], parts[2]);
    }

    public static bool IsNewer(string latest, string current)
    {
        var a = NormalizeVersion(latest);
        var b = NormalizeVersion(current);
        var cmp = a.Major.CompareTo(b.Major);
        if (cmp != 0) return cmp > 0;
        cmp = a.Minor.CompareTo(b.Minor);
        if (cmp != 0) return cmp > 0;
        return a.Patch.CompareTo(b.Patch) > 0;
    }

    private UpdateStatus SnapshotUnlocked() => new()
    {
        Current = _status.Current,
        Latest = _status.Latest,
        Available = _status.Available,
        DownloadUrl = _status.DownloadUrl,
        AssetName = _status.AssetName,
        Checked = _status.Checked,
        Error = _status.Error,
        Applying = _status.Applying,
    };

    private async Task<JsonDocument> DefaultFetchJsonAsync(string url)
    {
        using var req = new HttpRequestMessage(HttpMethod.Get, url);
        req.Headers.TryAddWithoutValidation("X-GitHub-Api-Version", "2022-11-28");
        using var resp = await _http.SendAsync(req).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();
        await using var stream = await resp.Content.ReadAsStreamAsync().ConfigureAwait(false);
        return await JsonDocument.ParseAsync(stream).ConfigureAwait(false);
    }

    private static bool MatchesPlatform(string name)
    {
        var lower = name.ToLowerInvariant();
        if (OperatingSystem.IsWindows())
            return lower.StartsWith("swiftsend-setup-", StringComparison.Ordinal) && lower.EndsWith(".exe", StringComparison.Ordinal);
        if (OperatingSystem.IsMacOS())
            return lower.Contains("macos", StringComparison.Ordinal) && lower.EndsWith(".dmg", StringComparison.Ordinal);
        return lower.Contains("linux", StringComparison.Ordinal) && lower.EndsWith(".appimage", StringComparison.Ordinal);
    }

    private async Task<string> DownloadAsync(string url, string assetName)
    {
        Directory.CreateDirectory(_downloadDir);
        var safeName = Path.GetFileName(assetName);
        var dest = Path.Combine(_downloadDir, safeName);
        using var resp = await _http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();
        await using var input = await resp.Content.ReadAsStreamAsync().ConfigureAwait(false);
        await using var output = File.Create(dest);
        await input.CopyToAsync(output).ConfigureAwait(false);
        if (new FileInfo(dest).Length <= 0)
            throw new InvalidOperationException("Download vazio");
        return dest;
    }

    private void SpawnApplyScript(string installer)
    {
        var scriptName = OperatingSystem.IsWindows() ? "apply_update.ps1" : "apply_update.sh";
        var src = Path.Combine(_scriptsDir, scriptName);
        if (!File.Exists(src))
        {
            var beside = Path.Combine(AppContext.BaseDirectory, "scripts", scriptName);
            if (File.Exists(beside))
                src = beside;
            else
                throw new FileNotFoundException($"Script de update não encontrado: {scriptName}");
        }

        var staging = Path.Combine(Path.GetTempPath(), scriptName);
        File.Copy(src, staging, overwrite: true);
        var pid = Environment.ProcessId;

        if (OperatingSystem.IsWindows())
        {
            // UseShellExecute=true: processo independente do job do app (sobrevive ao force kill).
            var psi = new ProcessStartInfo
            {
                FileName = "powershell.exe",
                Arguments =
                    $"-NoProfile -ExecutionPolicy Bypass -File \"{staging}\" -TargetPid {pid} -Installer \"{installer}\" -TimeoutSec {ForceKillTimeoutSec}",
                UseShellExecute = true,
                WindowStyle = ProcessWindowStyle.Hidden,
                WorkingDirectory = Path.GetTempPath(),
            };
            Process.Start(psi);
        }
        else
        {
            try
            {
                File.SetUnixFileMode(staging, UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute);
            }
            catch
            {
                // ignore on platforms without chmod semantics
            }

            var psi = new ProcessStartInfo
            {
                FileName = "/bin/bash",
                ArgumentList = { staging, pid.ToString(), installer, ForceKillTimeoutSec.ToString() },
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            Process.Start(psi);
        }
    }
}

File: python/update_service.py
"""Checagem e aplicação de atualizações via GitHub Releases."""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
import tempfile
import threading
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

GITHUB_REPO = "GuilhermeRoesler/SwiftSend"
GITHUB_API_LATEST = f"https://api.github.com/repos/{GITHUB_REPO}/releases/latest"
USER_AGENT = "SwiftSend-Updater"
CHECK_TIMEOUT_SEC = 12
DOWNLOAD_TIMEOUT_SEC = 600
FORCE_KILL_TIMEOUT_SEC = 10


@dataclass
class UpdateStatus:
    current: str
    latest: str | None = None
    available: bool = False
    download_url: str | None = None
    asset_name: str | None = None
    checked: bool = False
    error: str | None = None
    applying: bool = False

    def as_dict(self) -> dict:
        return {
            "current": self.current,
            "latest": self.latest,
            "available": self.available,
            "download_url": self.download_url,
            "asset_name": self.asset_name,
            "checked": self.checked,
            "error": self.error,
            "applying": self.applying,
        }


FetchJson = Callable[[str], dict]


def normalize_version(tag: str) -> tuple[int, int, int]:
    raw = (tag or "").strip()
    if raw.lower().startswith("v"):
        raw = raw[1:]
    core = raw.split("-", 1)[0].split("+", 1)[0]
    parts: list[int] = []
    for piece in core.split("."):
        try:
            parts.append(int(piece))
        except ValueError:
            parts.append(0)
    while len(parts) < 3:
        parts.append(0)
    return parts[0], parts[1], parts[2]


def is_newer(latest: str, current: str) -> bool:
    return normalize_version(latest) > normalize_version(current)


def platform_asset_matcher(system: str | None = None) -> Callable[[str], bool]:
    plat = (system or sys.platform).lower()

    def windows(name: str) -> bool:
        lower = name.lower()
        return lower.startswith("swiftsend-setup-") and lower.endswith(".exe")

    def linux(name: str) -> bool:
        lower = name.lower()
        return "linux" in lower and lower.endswith(".appimage")

    def macos(name: str) -> bool:
        lower = name.lower()
        return "macos" in lower and lower.endswith(".dmg")

    if plat.startswith("win"):
        return windows
    if plat == "darwin":
        return macos
    return linux


def default_fetch_json(url: str) -> dict:
    req = urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "User-Agent": USER_AGENT,
            "X-GitHub-Api-Version": "2022-11-28",
        },
    )
    with urllib.request.urlopen(req, timeout=CHECK_TIMEOUT_SEC) as resp:
        return json.loads(resp.read().decode("utf-8"))


def read_version(search_roots: list[Path]) -> str:
    for root in search_roots:
        candidate = root / "VERSION"
        if candidate.is_file():
            text = candidate.read_text(encoding="utf-8").strip()
            if text:
                return text
    return "0.0.0-dev"


class UpdateService:
    def __init__(
        self,
        current_version: str,
        scripts_dir: Path,
        *,
        fetch_json: FetchJson | None = None,
        download_dir: Path | None = None,
    ) -> None:
        self._lock = threading.Lock()
        self._status = UpdateStatus(current=current_version)
        self._scripts_dir = scripts_dir
        self._fetch_json = fetch_json or default_fetch_json
        self._download_dir = download_dir or Path(tempfile.gettempdir()) / "SwiftSendUpdates"
        self._matcher = platform_asset_matcher()

    @property
    def status(self) -> UpdateStatus:
        with self._lock:
            return UpdateStatus(**self._status.__dict__)

    def start_background_check(self) -> None:
        thread = threading.Thread(target=self.check_now, name="SwiftSendUpdateCheck", daemon=True)
        thread.start()

    def mark_skipped(self, reason: str | None = None) -> None:
        """Marca checagem concluída sem update (ex.: modo desenvolvimento)."""
        with self._lock:
            self._status.checked = True
            self._status.available = False
            self._status.latest = None
            self._status.download_url = None
            self._status.asset_name = None
            self._status.error = reason

    def check_now(self) -> UpdateStatus:
        try:
            data = self._fetch_json(GITHUB_API_LATEST)
            tag = str(data.get("tag_name") or "").strip()
            if not tag:
                raise ValueError("Release sem tag_name")

            assets = data.get("assets") or []
            download_url = None
            asset_name = None
            for asset in assets:
                name = str(asset.get("name") or "")
                url = str(asset.get("browser_download_url") or "")
                if name and url and self._matcher(name):
                    download_url = url
                    asset_name = name
                    break

            latest_display = tag[1:] if tag[:1].lower() == "v" else tag
            newer = is_newer(tag, self._status.current)
            available = newer and bool(download_url)
            with self._lock:
                self._status.latest = latest_display
                self._status.available = available
                self._status.download_url = download_url
                self._status.asset_name = asset_name
                self._status.checked = True
                self._status.error = (
                    "Sem artefato para esta plataforma" if newer and not download_url else None
                )
                return UpdateStatus(**self._status.__dict__)
        except Exception as exc:  # noqa: BLE001 — falha de rede/API não deve derrubar o app
            with self._lock:
                self._status.checked = True
                self._status.available = False
                self._status.error = str(exc)
                return UpdateStatus(**self._status.__dict__)

    def apply_update(self) -> dict:
        with self._lock:
            if self._status.applying:
                return {"success": False, "error": "Atualização já em andamento"}
            if not self._status.available or not self._status.download_url or not self._status.asset_name:
                return {"success": False, "error": "Nenhuma atualização disponível"}
            self._status.applying = True
            url = self._status.download_url
            asset_name = self._status.asset_name

        try:
            installer = self._download(url, asset_name)
            self._spawn_apply_script(installer)
            return {"success": True, "installer": str(installer)}
        except Exception as exc:  # noqa: BLE001
            with self._lock:
                self._status.applying = False
            return {"success": False, "error": str(exc)}

    def _download(self, url: str, asset_name: str) -> Path:
        self._download_dir.mkdir(parents=True, exist_ok=True)
        safe_name = Path(asset_name).name
        dest = self._download_dir / safe_name
        req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
        with urllib.request.urlopen(req, timeout=DOWNLOAD_TIMEOUT_SEC) as resp, open(dest, "wb") as out:
            shutil.copyfileobj(resp, out)
        if dest.stat().st_size <= 0:
            raise RuntimeError("Download vazio")
        return dest

    def _resolve_script(self) -> Path:
        if sys.platform == "win32":
            name = "apply_update.ps1"
        else:
            name = "apply_update.sh"
        candidates = [
            self._scripts_dir / name,
            Path(getattr(sys, "_MEIPASS", "")) / "scripts" / name,
            Path(__file__).resolve().parent.parent / "scripts" / name,
        ]
        for path in candidates:
            if path.is_file():
                return path
        raise FileNotFoundError(f"Script de update não encontrado: {name}")

    def _spawn_apply_script(self, installer: Path) -> None:
        src = self._resolve_script()
        # Copia para TEMP: em frozen o _MEIPASS some quando o processo morre.
        staging = Path(tempfile.gettempdir()) / src.name
        shutil.copy2(src, staging)
        if sys.platform != "win32":
            staging.chmod(staging.stat().st_mode | 0o111)

        pid = os.getpid()
        if sys.platform == "win32":
            args = [
                "powershell.exe",
                "-NoProfile",
                "-ExecutionPolicy",
                "Bypass",
                "-File",
                str(staging),
                "-TargetPid",
                str(pid),
                "-Installer",
                str(installer),
                "-TimeoutSec",
                str(FORCE_KILL_TIMEOUT_SEC),
            ]
            creationflags = 0
            creationflags |= getattr(subprocess, "DETACHED_PROCESS", 0x00000008)
            creationflags |= getattr(subprocess, "CREATE_NEW_PROCESS_GROUP", 0x00000200)
            creationflags |= getattr(subprocess, "CREATE_NO_WINDOW", 0x08000000)
            subprocess.Popen(
                args,
                close_fds=True,
                creationflags=creationflags,
                stdin=subprocess.DEVNULL,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
        else:
            subprocess.Popen(
                ["/bin/bash", str(staging), str(pid), str(installer), str(FORCE_KILL_TIMEOUT_SEC)],
                start_new_session=True,
                stdin=subprocess.DEVNULL,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                close_fds=True,
            )
